### PR TITLE
ProjectService_AddUserToProject

### PR DESCRIPTION
**Pull request by Inzilbeth**
Missing a RabbitMQ check logic from UserService if a user actually exists

### CODE REVIEW

File: DigitalOffice/DigitalOffice.sln [hunk 1]
```diff
@@ -5,35 +5,35 @@ VisualStudioVersion = 16.0.30204.135
 MinimumVisualStudioVersion = 10.0.40219.1
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Services", "Services", "{7A10A936-5C0A-476A-AFC4-C5CF8457BBE0}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "UserService", "Services\UserService\UserService.csproj", "{57E38E67-C667-41E5-95C2-98D2DC3E50EA}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "UserService", "D:\Git Repository\DigitalOffice\DigitalOffice\Services\UserService\UserService.csproj", "{57E38E67-C667-41E5-95C2-98D2DC3E50EA}"
```
Comment (KostinP) [inline DigitalOffice/DigitalOffice.sln]: Hm...

File: DigitalOffice/Services/ProjectService/Commands/AddUserToProjectCommand.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            this.repository = repository;
+            this.workerMapper = workerMapper;
+            this.managerMapper = managerMapper;
+        }
+
+        public bool Execute(AddUserToProjectRequest request)
+        {
+            validator.ValidateAndThrow(request);
+
+            //check if user actually exists in UserService here
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Commands/AddUserToProjectCommand.cs]: Hmm... Add TODO?

File: DigitalOffice/Services/ProjectService/Mappers/Interfaces/IMapper.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace ProjectService.Mappers.Interfaces
+{
+    public interface IMapper<Tin, Tout>
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Mappers/Interfaces/IMapper.cs]: TIn and TOut

File: DigitalOffice/Services/ProjectService/Mappers/ManagerMapper.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace ProjectService.Mappers
+{
+    public class ManagerMapper : IMapper<AddUserToProjectRequest, DbProjectManagerUser>
+    {
+        public DbProjectManagerUser Map(AddUserToProjectRequest value)
+        {
+            if (value == null)
+            {
+                throw new ArgumentNullException("value");
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Mappers/ManagerMapper.cs]: Hmm... nameof?

File: DigitalOffice/Services/ProjectService/Mappers/WorkerMapper.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace ProjectService.Mappers
+{
+    public class WorkerMapper : IMapper<AddUserToProjectRequest, DbProjectWorkerUser>
+    {
+        public DbProjectWorkerUser Map(AddUserToProjectRequest value)
+        {
+            if (value == null)
+            {
+                throw new ArgumentNullException("value");
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Mappers/WorkerMapper.cs]: Hm.... nameof?

File: DigitalOffice/Services/ProjectService/Validators/AddUserToProjectRequestValidator.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace ProjectService.Validators
+{
+    public class AddUserToProjectRequestValidator : AbstractValidator<AddUserToProjectRequest>
+    {
+        public AddUserToProjectRequestValidator()
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Validators/AddUserToProjectRequestValidator.cs]: Spaces?

**Review by KostinP — CHANGES_REQUESTED** [review]
Resolve conflicts

File: DigitalOffice/Services/ProjectService/Controllers/ProjectController.cs [hunk 7]
```diff
@@ -1,4 +1,28 @@
 ﻿using Microsoft.AspNetCore.Mvc;
+<<<<<<< HEAD
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Controllers/ProjectController.cs]: Resolve conflict. Write to me if you need help with this, I'll show you how it's done.

File: DigitalOffice/Services/ProjectService/Mappers/Interfaces/IMapper.cs [hunk 8]
```diff
@@ -1,7 +1,20 @@
...
+    {
+        Tout Map(Tin value);
+=======
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Mappers/Interfaces/IMapper.cs]: Resolve conflict.

File: DigitalOffice/Services/ProjectService/Models/Project.cs [hunk 9]
```diff
@@ -1,12 +1,28 @@
 ﻿using System;
 using System.Collections.Generic;
 
+<<<<<<< HEAD
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Models/Project.cs]: Resolve conflict.

File: DigitalOffice/Services/ProjectService/Repositories/Interfaces/IProjectRepository.cs [hunk 10]
```diff
@@ -1,3 +1,19 @@
+<<<<<<< HEAD
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Repositories/Interfaces/IProjectRepository.cs]: Resolve conflict.

File: DigitalOffice/Services/ProjectService/Repositories/ProjectRepository.cs [hunk 11]
```diff
@@ -1,3 +1,19 @@
...
+using ProjectService.Models;
+using ProjectService.Repositories.Interfaces;
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace ProjectService.Repositories
+=======
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Repositories/ProjectRepository.cs]: Resolve conflict.

File: DigitalOffice/Services/ProjectService/Commands/Interfaces/IAddUserToProjectCommand.cs [hunk 12]
```diff
@@ -0,0 +1,9 @@
+﻿using ProjectService.Models;
+
+namespace ProjectService.Commands.Interfaces
+{
+    public interface IAddUserToProjectCommand
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Commands/Interfaces/IAddUserToProjectCommand.cs]: Add XML documentation.

File: DigitalOffice/Services/ProjectService/Mappers/Interfaces/IMapper.cs (hunk 3 above)
Comment (KostinP) [reply DigitalOffice/Services/ProjectService/Mappers/Interfaces/IMapper.cs]: Add XML documentation.

File: DigitalOffice/Services/ProjectService/Repositories/Interfaces/IProjectRepository.cs [hunk 13]
```diff
@@ -1,3 +1,19 @@
...
+using ProjectService.Database.Entities;
+using ProjectService.Models;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace ProjectService.Repositories.Interfaces
+{
+    public interface IProjectRepository
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Repositories/Interfaces/IProjectRepository.cs]: Add XML documentation.

**Review by KostinP — CHANGES_REQUESTED** [review]

File: DigitalOffice/Services/ProjectService/Broker/Requests/IUserExistenceRequest.cs [hunk 14]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+
+namespace LT.DigitalOffice.ProjectService.Broker.Requests
+{
+    public interface IUserExistenceRequest
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Broker/Requests/IUserExistenceRequest.cs]: Add XML

File: DigitalOffice/Services/ProjectService/Broker/Responses/IUserExistenceResponse.cs [hunk 15]
```diff
@@ -0,0 +1,7 @@
+﻿namespace LT.DigitalOffice.ProjectService.Broker.Responses
+{
+    public interface IUserExistenceResponse
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Broker/Responses/IUserExistenceResponse.cs]: Add XML

File: DigitalOffice/Services/ProjectService/Broker/Senders/Interfaces/ISender.cs [hunk 16]
```diff
@@ -0,0 +1,11 @@
+﻿using LT.DigitalOffice.Kernel.Broker;
+using MassTransit;
+using System.Threading.Tasks;
+
+namespace LT.DigitalOffice.ProjectService.Broker.Senders.Interfaces
+{
+    public interface ISender<TIn, TOut>
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Broker/Senders/Interfaces/ISender.cs]: Add XML

File: DigitalOffice/Services/ProjectService/Commands/Interfaces/IAddUserToProjectCommand.cs (hunk 12 above)
Comment (KostinP) [reply DigitalOffice/Services/ProjectService/Commands/Interfaces/IAddUserToProjectCommand.cs]: Fixed.

File: DigitalOffice/Services/ProjectService/Commands/Interfaces/IAddUserToProjectCommand.cs [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+    /// <summary>
+    /// Represents interface for a command in command pattern.
+    /// Provides method for adding a new user to a project.
+    /// </summary>
+    public interface IAddUserToProjectCommand
+    {
+        /// <summary>
+        /// Adds a new user to a project. 
+        /// </summary>
+        /// <param name="request">Request containing the user id & project id.</param>
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Commands/Interfaces/IAddUserToProjectCommand.cs]: Replace from & to and

**Review by KostinP — CHANGES_REQUESTED** [review]

File: DigitalOffice/Services/ProjectService/Broker/Requests/IUserExistenceRequest.cs [hunk 18]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+
+namespace LT.DigitalOffice.ProjectService.Broker.Requests
+{
+    /// <summary>
+    /// Iterface for a request that is sent to check User's existence in USerService.
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Broker/Requests/IUserExistenceRequest.cs]: Change USerService to UserService

File: DigitalOffice/Services/ProjectService/Controllers/ProjectController.cs [hunk 19]
```diff
@@ -19,5 +20,14 @@ public Project GetProjectInfoById([FromServices] IGetProjectInfoByIdCommand comm
         public Guid CreateNewProject(
             [FromServices] ICreateNewProjectCommand command,
             [FromBody] NewProjectRequest request) => command.Execute(request);
+
+        [HttpPut]
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Controllers/ProjectController.cs]: Why is there no route?

File: DigitalOffice/Services/ProjectService/Validators/AddUserToProjectRequestValidator.cs (hunk 6 above)
Comment (KostinP) [reply DigitalOffice/Services/ProjectService/Validators/AddUserToProjectRequestValidator.cs]: I will write in more detail. It seems that we will separate the rules with empty lines everywhere

File: DigitalOffice/Services/ProjectService/Validators/AddUserToProjectRequestValidator.cs [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+        {
+            RuleFor(r => r.UserId)
+                .NotEmpty()
+                .NotNull();
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Validators/AddUserToProjectRequestValidator.cs]: Why NotEmpty and NotNull?

File: DigitalOffice/Services/ProjectService/Validators/AddUserToProjectRequestValidator.cs [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+            RuleFor(r => r.ProjectId)
+                .NotEmpty()
+                .NotNull();
```
Comment (KostinP) [inline DigitalOffice/Services/ProjectService/Validators/AddUserToProjectRequestValidator.cs]: Why NotEmpty and NotNull?

File: DigitalOffice/Services/UserService/Broker/Requests/IUserExistenceRequest.cs [hunk 22]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+
+namespace LT.DigitalOffice.UserService.Broker.Requests
+{
+    /// <summary>
+    /// Iterface for a request that is sent to check User's existence in USerService.
```
Comment (KostinP) [inline DigitalOffice/Services/UserService/Broker/Requests/IUserExistenceRequest.cs]: UserService*

File: DigitalOffice/Tests/ProjectServiceTests/Commands/AddUserToProjectCommandTests.cs [hunk 23]
```diff
@@ -0,0 +1,210 @@
...
+            public Uri DestinationAddress { get; }
+
+            public Uri ResponseAddress { get; }
+
+            public Uri FaultAddress { get; }
+
+            public DateTime? SentTime { get; }
+
+            public Headers Headers { get; }
+
```
Comment (KostinP) [inline DigitalOffice/Tests/ProjectServiceTests/Commands/AddUserToProjectCommandTests.cs]: Why are there so many blank lines?

**Review by KostinP — CHANGES_REQUESTED** [review]

File: DigitalOffice/Services/UserService/Broker/Requests/IUserExistenceRequest.cs (hunk 22 above)
Comment (MihailKulikov) [reply DigitalOffice/Services/UserService/Broker/Requests/IUserExistenceRequest.cs]: Iterface

File: DigitalOffice/IntrnalModels/TemporaryIntrnalModelsLib.csproj [hunk 24]
```diff
@@ -0,0 +1,7 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (MihailKulikov) [inline DigitalOffice/IntrnalModels/TemporaryIntrnalModelsLib.csproj]: ???

File: DigitalOffice/Services/ProjectService/Broker/Requests/UserExistenceRequest.cs [hunk 25]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+
+namespace LT.DigitalOffice.ProjectService.Broker.Requests
+{
+    public class UserExistenceRequest : IUserExistenceRequest
```
Comment (MihailKulikov) [inline DigitalOffice/Services/ProjectService/Broker/Requests/UserExistenceRequest.cs]: What is the purpose of the class?

File: DigitalOffice/Services/ProjectService/Broker/Responses/UserExistenceResponse.cs [hunk 26]
```diff
@@ -0,0 +1,7 @@
+﻿namespace LT.DigitalOffice.ProjectService.Broker.Responses
+{
+    public class UserExistenceResponse
```
Comment (MihailKulikov) [inline DigitalOffice/Services/ProjectService/Broker/Responses/UserExistenceResponse.cs]: ???

File: DigitalOffice/Services/ProjectService/Broker/Responses/IUserExistenceResponse.cs [hunk 27]
```diff
@@ -0,0 +1,10 @@
+﻿namespace LT.DigitalOffice.ProjectService.Broker.Responses
+{
+    /// <summary>
+    /// Interface for a respond that shows whether the user exists or not.
+    /// </summary>
+    public interface IUserExistenceResponse
+    {
+        bool Exists { get; }
```
Comment (MihailKulikov) [inline DigitalOffice/Services/ProjectService/Broker/Responses/IUserExistenceResponse.cs]: Why not just use a boolean variable in IOperationResult?

File: DigitalOffice/Services/ProjectService/Commands/AddUserToProjectCommand.cs [hunk 28]
```diff
@@ -0,0 +1,49 @@
...
+            this.repository = repository;
+            this.projectUserMapper = projectUserMapper;
+            this.userExistenceSender = userExistenceSender;
+        }
+
+        public async Task<bool> Execute(AddUserToProjectRequest request)
+        {
+            validator.ValidateAndThrow(request);
+
+            // TODO: check if user actually exists in UserService here
```
Comment (MihailKulikov) [inline DigitalOffice/Services/ProjectService/Commands/AddUserToProjectCommand.cs]: Remove

File: DigitalOffice/Services/ProjectService/Mappers/ProjectUserMapper.cs [hunk 29]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace LT.DigitalOffice.ProjectService.Mappers
+{
+    public class ProjectUserMapper : IMapper<AddUserToProjectRequest, DbProjectWorkerUser>
+    {
+        public DbProjectWorkerUser Map(AddUserToProjectRequest value)
+        {
+            if (value == null)
+            {
+                throw new ArgumentNullException("value");
```
Comment (MihailKulikov) [inline DigitalOffice/Services/ProjectService/Mappers/ProjectUserMapper.cs]: nameof?

File: DigitalOffice/Services/ProjectService/Models/AdduserToProjectRequest.cs [hunk 30]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace LT.DigitalOffice.ProjectService.Models
+{
+    public class AddUserToProjectRequest
```
Comment (MihailKulikov) [inline DigitalOffice/Services/ProjectService/Models/AdduserToProjectRequest.cs:5]: Useless

File: DigitalOffice/Services/UserService/Broker/Requests/UserExistenceRequest.cs [hunk 31]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+
+namespace LT.DigitalOffice.UserService.Broker.Requests
+{
+    public class UserExistenceRequest : IUserExistenceRequest
```
Comment (MihailKulikov) [inline DigitalOffice/Services/UserService/Broker/Requests/UserExistenceRequest.cs]: ???

File: DigitalOffice/Services/UserService/Broker/Responses/UserExistenceResponse.cs [hunk 32]
```diff
@@ -0,0 +1,7 @@
+﻿namespace LT.DigitalOffice.UserService.Broker.Responses
+{
+    public class UserExistenceResponse
```
Comment (MihailKulikov) [inline DigitalOffice/Services/UserService/Broker/Responses/UserExistenceResponse.cs]: ???

**Review by MihailKulikov — CHANGES_REQUESTED** [review]
Add tests for UserExistenceConsumer,, UserExistenceSender, ProjectController

File: DigitalOffice/Services/ProjectService/Broker/Responses/IUserExistenceResponse.cs (hunk 27 above)
Comment (Inzilbeth) [reply DigitalOffice/Services/ProjectService/Broker/Responses/IUserExistenceResponse.cs]: I wasn't sure if we would want to modify the response in the future or not

File: DigitalOffice/Services/ProjectService/Models/AdduserToProjectRequest.cs (hunk 30 above)
Comment (Inzilbeth) [inline DigitalOffice/Services/ProjectService/Models/AdduserToProjectRequest.cs]: What do you mean?

File: DigitalOffice/Services/ProjectService/Broker/Senders/UserExistenceSender.cs [hunk 33]
```diff
@@ -0,0 +1,42 @@
...
+using LT.DigitalOffice.ProjectService.Broker.Senders.Interfaces;
+using MassTransit;
+using Microsoft.AspNetCore.Mvc;
+using System;
+using System.Threading.Tasks;
+
+namespace LT.DigitalOffice.ProjectService.Broker.Senders
+{
+    public class UserExistenceSender : ISender<Guid, IUserExistenceResponse>
+    {
```
Comment (djino97) [inline DigitalOffice/Services/ProjectService/Broker/Senders/UserExistenceSender.cs]: remove, move broker logic into command

File: DigitalOffice/Services/ProjectService/Broker/Senders/Interfaces/ISender.cs [hunk 34]
```diff
@@ -0,0 +1,16 @@
...
+using System.Threading.Tasks;
+
+namespace LT.DigitalOffice.ProjectService.Broker.Senders.Interfaces
+{
+    /// <summary>
+    /// Broker sender interface.
+    /// </summary>
+    /// <typeparam name="TIn">Input type.</typeparam>
+    /// <typeparam name="TOut">Output type.</typeparam>
+    public interface ISender<TIn, TOut>
```
Comment (djino97) [inline DigitalOffice/Services/ProjectService/Broker/Senders/Interfaces/ISender.cs]: remove interface

File: DigitalOffice/Services/UserService/Broker/Consumers/UserExistenceConsumer.cs [hunk 35]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            var response = new 
+            {
+                IsSuccess = true,
+                Errors = new List<string>(),
+                Body = new { Exists = repository.UserExists(context.Message.Id) }
+            };
+
+            await context.RespondAsync<IOperationResult<IUserExistenceResponse>>(response);
+        }
```
Comment (djino97) [inline DigitalOffice/Services/UserService/Broker/Consumers/UserExistenceConsumer.cs]: Use wrapper OperationResult

File: Tests/ProjectServiceTests/Controllers/ProjectControllerTests.cs [hunk 36]
```diff
@@ -0,0 +1,36 @@
...
+    class ProjectControllerTests
+    {
+        private ProjectController projectController;
+        private Mock<IAddUsersToProjectCommand> command;
+
+        [SetUp]
+        public void Setup()
+        {
+            projectController = new ProjectController();
+
```
Comment (KostinP) [inline Tests/ProjectServiceTests/Controllers/ProjectControllerTests.cs:20]: Extra empty line.

File: Tests/ProjectServiceTests/Senders/UserExistenceSenderTests.cs [hunk 37]
```diff
@@ -0,0 +1,7 @@
+﻿namespace LT.DigitalOffice.ProjectServiceUnitTests.Senders
+{
+    class UserExistenceSenderTests
+    {
+
```
Comment (KostinP) [inline Tests/ProjectServiceTests/Senders/UserExistenceSenderTests.cs:5]: ?

File: Tests/UserServiceTests/Consumers/UserExistenceConsumerTests.cs [hunk 38]
```diff
@@ -0,0 +1,6 @@
+﻿namespace LT.DigitalOffice.UserServiceUnitTests.Consumers
+{
+    class UserExistenceConsumerTests
+    {
```
Comment (KostinP) [inline Tests/UserServiceTests/Consumers/UserExistenceConsumerTests.cs:4]: ?

File: DigitalOffice/DigitalOffice.sln (hunk 1 above)
Comment (DarkFelas) [reply DigitalOffice/DigitalOffice.sln]: Return old code

File: DigitalOffice/Services/ProjectService/Broker/Responses/IUserExistenceResponse.cs (hunk 27 above)
Comment (DarkFelas) [reply DigitalOffice/Services/ProjectService/Broker/Responses/IUserExistenceResponse.cs]: Use IOperationResult

**Pull request closed without merging**